pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,395 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tcpproxy
+
+import (
+	"errors"
+	"io"
+	"net"
+	"runtime/debug"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/util/iobufferpool"
+	"github.com/megaease/easegress/pkg/util/limitlistener"
+	"github.com/megaease/easegress/pkg/util/timerpool"
+)
+
+const writeBufSize = 8
+
+var tcpBufferPool = sync.Pool{
+	New: func() interface{} {
+		return make([]byte, iobufferpool.DefaultBufferReadCapacity)
+	},
+}
+
+// Connection wrap tcp connection
+type Connection struct {
+	closed     uint32
+	rawConn    net.Conn
+	localAddr  net.Addr
+	remoteAddr net.Addr
+
+	lastBytesSizeRead  int
+	lastWriteSizeWrite int
+
+	readBuffer      []byte
+	writeBuffers    net.Buffers
+	ioBuffers       []*iobufferpool.StreamBuffer
+	writeBufferChan chan *iobufferpool.StreamBuffer
+
+	mu               sync.Mutex
+	startOnce        sync.Once
```

Comment:
why we need to use sync.Once here? do we expect Start to be called multiple times and concurrently?